app/src/main/java/io/github/sds100/keymapper/actions/pinchscreen/PinchPickDisplayCoordinateViewModel.kt

```diff
@@ -0,0 +1,240 @@
+package io.github.sds100.keymapper.actions.pinchscreen
+
+import android.graphics.Bitmap
+import android.graphics.Point
+import android.view.View
+import android.widget.AdapterView
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import io.github.sds100.keymapper.R
+import io.github.sds100.keymapper.util.ui.*
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
+import kotlin.math.roundToInt
+
+class PinchPickDisplayCoordinateViewModel(
+    resourceProvider: ResourceProvider
+) : ViewModel(), ResourceProvider by resourceProvider, PopupViewModel by PopupViewModelImpl() {
+
+    private val pinchTypes = arrayOf(PinchScreenType.PINCH_IN.name, PinchScreenType.PINCH_OUT.name)
+
+    private val x = MutableStateFlow<Int?>(null)
+    private val y = MutableStateFlow<Int?>(null)
+    private val distance = MutableStateFlow<Int?>(null)
+    private val pinchType = MutableStateFlow<PinchScreenType?>(PinchScreenType.PINCH_IN)
+    private val fingerCount = MutableStateFlow<Int?>(2)
+    private val duration = MutableStateFlow<Int?>(200)
+
+    private val _bitmap = MutableStateFlow<Bitmap?>(null)
+    private val _returnResult = MutableSharedFlow<PinchPickCoordinateResult>()
+
+    private val description: MutableStateFlow<String?> = MutableStateFlow(null)
+
+    val xString = x.map {
+        it ?: return@map ""
+
+        it.toString()
+    }.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    val yString = y.map {
+        it ?: return@map ""
+
+        it.toString()
+    }.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    val distanceString = distance.map {
+        it ?: return@map ""
+
+        it.toString()
+    }.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    val pinchTypeSpinnerSelection = pinchType.map {
+        it ?: return@map 0
+
+        this.pinchTypes.indexOf(it.name)
+    }.stateIn(viewModelScope, SharingStarted.Lazily, 0)
+
+    val fingerCountString = fingerCount.map {
+        it ?: return@map ""
+
+        it.toString()
+    }.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    val fingerCountError: StateFlow<String?> = fingerCount.map { count ->
+        if (count == null) {
+            return@map resourceProvider.getString(R.string.error_cant_be_empty)
+        }
+
+        if (count < 2) {
+            return@map resourceProvider.getString(R.string.error_pinch_screen_must_be_two_or_more_fingers)
+        }
+
+        if (count > 10) {
+            return@map resourceProvider.getString(R.string.error_pinch_screen_must_be_ten_or_less_fingers)
+        }
+
+        null
+    }.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    val durationString = duration.map {
+        it ?: return@map ""
+
+        it.toString()
+    }.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    val durationError: StateFlow<String?> = duration.map { d ->
```

Comment:
can you do the same for `SwipePickDisplayCoordinateViewModel`

Reply:
done 👍